File: Python/Truck.py
from abc import ABC, abstractmethod

class Truck(ABC):
    truck_hours = 0
    totalVolume = 0
    totalWeight = 0
    packages = []
    count = 1

    def __init__(self):
        self.truckID = Truck.count
        Truck.count += 1
        self.packages = []

    @abstractmethod
    def add_pack(self, pack):
        self.packages.append(pack)

    def remove_pack(self, pack):
        self.packages.remove(pack)

    def get_packages(self):
        return self.packages

    def get_truckID(self):
        return self.truckID

    @abstractmethod
    def total_volume(self):
        for pack in self.packages:
            self.totalVolume += pack.get_volume()
        return self.totalVolume

    @abstractmethod
    def total_weight(self):
        for pack in self.packages:
            self.totalWeight += pack.get_weight()
        return self.totalWeight

    def total_truck_hours(self):
        for pack in self.packages:
            self.truck_hours += pack.get_time()
        return self.truck_hours

    def __str__(self):
        return f"Truck ID: {self.get_truckID()} Total Volume: {self.total_volume()} Total Weight: {self.total_weight()}"

File: CPP/Truck.cpp
#include "headers/Truck.h"
#include <string>

int Truck::count = 0;

// default constructor
Truck::Truck() {
    truckID = count;
    count++;
}

// accessor methods
CustomList<Pack> Truck::getPackages() const {
    return packages;
}

int Truck::getTruckID() const {
    return truckID;
}

// check back on this after implementing SpecialPack
int Truck::getTruckHours() {
    for (int i=0; i < packages.size(); i++) {
        truckHours += Pack::getTime();
    }
    return truckHours;
}

int Truck::getMaxWeight() const {
    return MAX_WEIGHT;
}

int Truck::getMaxVolume() const  {
    return MAX_VOLUME;
}

// other methods
void Truck::addPack(const Pack& p) {
    packages.add(p);
}

void Truck::removePack(const Pack& p) {
    packages.remove(p);
}

int Truck::totalPacksVolume() {
    for (int i=0; i < packages.size(); i++) {
        totalVolume += packages.get(i).getVolume();
    }
    return totalVolume;
}

int Truck::totalPacksWeight() {
   for (int i=0; i < packages.size(); i++) {
       totalWeight += packages.get(i).getWeight();
   }
   return totalWeight;
}

// check back on this after implementing SpecialPack
int Truck::totalTruckHours() const {
    int truckHrs = 0;
    for (int i=0; i < packages.size(); i++) {
        truckHrs += Pack::getTime();
    }
    return truckHrs;
}

std::string Truck::toString() {
    return "Truck ID: " + std::to_string(getTruckID()) + " Total Volume: " + std::to_string(totalPacksVolume()) +
    " Total Weight: " + std::to_string(totalPacksWeight()) ;
}

